Comrv master

ronen-haen-sndk opened the pull request.

ronen-haen-sndk committed:
Comrv baremetal (#10)

* fix build for rtos demo

* add comrv entry implementation

* use a register save caller ra offset and token

* remove License directory

* fix build warning

* make comrv compile with llvm

* add COM-RV stack support

* add comrv stack support

* fix build

* backup

* backup - first real comrv call works

* move offset calculation from .S to .c file + add main implementation

* add comments

* linker script- change .ovlgrp.x to .ovlgrpdata
comrv.c - modify comments, rename some macros and handle return offset with multi group
comrv_entry.S - add overlay support with function pointer (thunks)
SConstruct - move here the -ffixed-regX option

* code review changes
- make comrvGetAddressFromToken to receive no args
- use D_COMRV_INVOKE_CALLEE_BIT_0 instead of D_COMRV_PROFILING_INVOKE_VAL and D_COMRV_RET_CALLER_BIT_0
- +additional minor changes

* add missing thunk field to comrvTokenFields_t

*  changes following CR

* add comment to hook functions

* remove NULL #define

* changes following CR

* CR changes:
1. add comrvGetStatus function
2. in comrvInit, move pStackStartAddr->calleeToken = 0; to be done only for the last item of the stack

* CR changes:
1. in comrvInit remove duplication of pStackStartAddr->offsetPrevFrame = D_COMRV_END_OF_STACK (initially it was done in the wrong frame)

* CR changes:
1. add comrvInitApplicationStack function to initialize comrv application stack

* remove volatile'

* CR changes:
1. remove comrv stack from ld script
2. add argument to comrvInit
3. make the comrv stack and comrv CB globals instead of receiving them from ld
4. add build error in case no eviction algorithm selected
5. add comments + vars naming fixes

* CR changes:
1. add empty macros for enter/exit critical section

* CR changes:
1. add comments
2. stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress
assign local meaningful arg
3. cEvictCandidateList[ucEntryIndex] should be ucEvictCandidateList[ucEntryIndex+1]

* CR changes:
1. rename remaining 'heap' to 'cache'
2. modify lru fields to be union
3. in comrvInit decrease code size by using local variable pCacheEntry
4. in comrvUpdateCacheEntryAccess make code more readable by adding local pointer to point to the updated entry (&stComrvCB.stOverlayCache[ucEntryIndex])

* add comment to usSelectedMultiGroupEntry

* CR changes:
1. remove unused #define
2. reorder condition check for ucNumOfEvictionCandidates

* CR changes:
1. rename comrvUpdateCacheEntryAccess to comrvUpdateCacheEvectionParams
2. fix properties size field in case the evicted size was bigger than the requested size
3. comments

* CR changes:
remane typNextIndex to typNextLruIndex and typPrevIndex to typPrevLruIndex

* CR changes:
1. comments
2. add macro to convert group size to bytes

* CR changes:
1. change comrvLoadOvlayGroupHook() to recieve structure instead of several args

*  CR changes:
1. in comrvGetAddressFromToken remove (void*) casting in return value

*  CR changes:
1. clear the usCalleeMultiGroupTableEntry field so it won't be reused
2. fix typo

* CR changes:
1. rename stComrvCB to g_stComrvCB
2. rename stComrvStackPool to g_stComrvStackPool

* CR changes:
1. change D_COMRV_EVICT_CANDIDATE_MAP_SIZE to use the values from the comrv_config.h file
2. add comrv version memcpy instead of using the library function

* CR changes:
in comrvGetEvictionCandidates add a local variable to hold stComrvCB.stOverlayCache[ucEntryIndex]

* fix build issue

* CR changes:
1. remove comrvNotificationHook
2. add comrvInstrumentationHook
3. add comrvErrorHook
4. fix instrumentation #define names

* CR changes:
use comrv macro for inline

* CR changes:
1. change D_COMRV_LRU_LAST_ITEM to D_COMRV_MRU_ITEM
2. change D_COMRV_LRU_FIRST_ITEM to D_COMRV_LRU_ITEM

* fix wrong arg passing to comrvErrorHook

* in comrvSearchForLoadedOverlayGroup fix loop counting

* CR changes:
1. fix overlay data condition check in comrvGetAddressFromToken
2. change comrvSearchForLoadedOverlayGroup to return an index instead of an address

* CR changes:
in comrvGetEvictionCandidates, make the code more readble in the loop 'while (uiCandidates)'

* CR changes:
1. rename __OVERLAY_GROUPS_START__ to __OVERLAY_STORAGE_START__ADDRESS__
2. remove unnecessary #include
3. fix comments and typos

* CR changes:
in comrvGetEvictionCandidates add a check if the  accumulated size < requestedsize and call comrvErrorHook accordingly

* fix 'Load address misaligned' exception

* CR changes:
invoke CRC via macro

* fix a potential bug (due to latest changes) when returning from comrvGetEvictionCandidates we need to make sure we exit the critical section so we can't call directly to comrvErrorHook from comrvGetEvictionCandidates

* CR changes:
in comrvGetEvictionCandidates assert if we reached a case where we have too many entries available

* fix comment

* CR changes:
add to comrvErrorHook call to comrvGetStatus

* CR changes:
echo error if llvm isn't used to compile comrv module

* CR changes:
1. fix D_COMRV_NUM_OF_CACHE_ENTRIES definition - not by the end user
2. add D_COMRV_EVICTION_POLICY definition - defined by the end user
3. typo

* CR changes:rename PROFILE -> INSTRUMENTATION

* CR changes:
1. move all types to comrv_types.h
2. move all #defines to comrv_defines.h
3. in comrvStackFrame_t add one padding of u16_t

*  CR changes:
rename comrv_asm_macro.h -> comrv_macros.inc

* CR changes:
1. overlay cache - use linker symbols instead of global variables
2. overlay tables - use linker symbols instead of global variables
3. fix build issue due to rename to comrv_macros.inc

* CR changes:
1. delete pFixedEntryAddress field from comrvCacheEntry_t; instead use linker symbols to calculate the address of the accessed cache entry
2. in main, add code to assist in future debug - set the mtvec register so we'll be able to break in case of exception

* CR changes:
use __builtin_expect in errors to avoid branching

* remove comment

* comments

* Fix potential bug due to wrong change I made in fragmentation:
1. when handling fragmentation I used the wrong index
2. make M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES use M_COMRV_GET_OVL_GROUP_SIZE
3. comments

* remove unused local variable in comrvInit

* CR changes:
add comment

* CR changes:
add missing prefix

ronen-haen-sndk committed:
align comrv to latest build environment

ronen-haen-sndk committed:
merge with master

ronen-haen-sndk committed:
remove old code

ronen-haen-sndk committed:
add support for objcopy

ronen-haen-sndk committed:
remove old files

ronen-haen-sndk committed:
fix post build actions issue with scons

ronen-haen-sndk committed:
remove old code

ronen-haen-sndk committed:
add missing license header

ronen-haen-sndk committed:
Comrv integ llvm (#20)

* add implementation - first overlay group contains comrv tables:
add comrvLoadTables()
add init args field
in comrvInit add conditional call to comrvLoadTables() - depend on init args
remove TODO comments
remove permanent init BP in hifive1.launch

* align to new build tools:
fix address of comrv tables
align linker script
fix search loop counter
fix grouping file to start from group 1
renane -foverlay-functions to -fcomrv and remove all -ffixed-reg

* remove the noinline from __OVERLAY__ define

* minor script change

* multi group support - first working demo

* remove redundant code - reduce code size

ronen-haen-sndk committed:
reproduce auipc use in overlay function

ronen-haen-sndk committed:
fine tune the use of fence.i - to make sure the overlay group is loaded before call/return

ronen-haen-sndk committed:
disable warning 'Instruction modifies reserved register'

ronen-haen-sndk committed:
fix thunk handling

oferShinaar commented on WD-Firmware/board/hifive1/link.lds:
```diff
@@ -48,6 +53,12 @@ SECTIONS
     *(.gnu.linkonce.r.*)
   } >flash AT>flash :flash
 
+  .reserved_ovl  :
```
please add header comment

ronen-haen-sndk replied:
done in 8dce911

oferShinaar commented on WD-Firmware/board/hifive1/link.lds:
```diff
@@ -152,10 +163,27 @@ SECTIONS
     PROVIDE( _sp = . );
   } >ram AT>ram :ram
+
+  .overlay_sec ALIGN(__comrv_align_size) :
```
please add header comment

ronen-haen-sndk replied:
done in 8dce911

oferShinaar commented on WD-Firmware/board/hifive1/link.lds:
```diff
@@ -152,10 +163,27 @@ SECTIONS
+  }  >ram AT>ram :ram
+
+  /* Overlay data */
+  .ovlgrpdata       : 
```
please add header comment

ronen-haen-sndk replied:
done in 8dce911

oferShinaar commented on WD-Firmware/comrv/api_inc/comrv_defines.h:
```diff
@@ -27,9 +27,90 @@
 * definitions
 */
+// TODO: ronen - use psp defines
+#define D_COMRV_INLINE     __attribute__((inline))
```
Consider: if the use must be "inline" then there is "always inline" attribute.

ronen-haen-sndk replied:
done in f224b36

oferShinaar commented on WD-Firmware/comrv/comrv.c:
```diff
@@ -18,21 +18,113 @@
+* @brief  The file implements the COM-RV engine
 * 
 */
+_Pragma("clang diagnostic ignored \"-Winline-asm\"")
```
please add comment for why we are doing this

ronen-haen-sndk replied:
done in da336bd

oferShinaar commented on WD-Firmware/comrv/loc_inc/comrv.h:
```diff
@@ -35,6 +35,10 @@
+#define M_COMRV_ENTER_CRITICAL_SECTION()
+#define M_COMRV_EXIT_CRITICAL_SECTION()
+// TODO: use our psp
+#define _BUILTIN_EXPECT(condition, expected)  __builtin_expect(condition, expected)
```
please add a comment why we are doing this

ronen-haen-sndk replied:
done in b4d5e7b

oferShinaar commented on WD-Firmware/demo/build/SConscript_comrv_baremetal:
```diff
@@ -46,6 +45,11 @@ listAsimCompilerDirectivs = [] + Env['A_FLAGS']
 
 # compilation defines (-D_)
 listCompilationDefines = [] + Env['PUBLIC_DEF']
+'''
```
can we take all ComRV "configurations" to a file ? (D_COMRV_EVICTION_POLICY, D_COMRV_VERIFY_INIT_ARGS,etc ....)
to not be part of the build system.
this way you have file with default values and user can changed this file if needed. keep in mind this is a lib that will not be compiled all the time.

ronen-haen-sndk replied:
done in 09b4fb8

oferShinaar commented on WD-Firmware/demo/build/SConstruct:
```diff
@@ -154,10 +142,12 @@ print "Building SConscripts files:\n\n-%s" % "\n-".join(objDemo.listSconscripts)
+
+objBundle = Env.Program(Env['ELF_FILE'], arrLibs, LIBS=Env['TARGET_LIBS'])
+# execute post build actions
+AddPostAction(objBundle, Action(utils.fnProduceSectionsSize))
```
we might find a better way to do this with depend. DONT change anything! just add a comment for "todo:" to check depend soultion

ronen-haen-sndk replied:
done in 2a82cc9

oferShinaar commented on WD-Firmware/demo/build/demos/config.py:
```diff
@@ -67,41 +83,53 @@ def getConfigure(self):
       exit(1)
     # read the configure file from he build folder and grab its info
```
please add one line space

ronen-haen-sndk replied:
done in ce738f1

oferShinaar commented on WD-Firmware/demo/build/demos/demo_comrv_baremetal.py:
```diff
@@ -27,8 +32,18 @@ def __init__(self):
+    ]
+
+    self.listDemoSpecificCFlags = [
+      '-fcomrv',
```
add comment to explain -fcomrv. this is user flag, we want them to understand

ronen-haen-sndk replied:
done in caec55c

oferShinaar commented on WD-Firmware/demo/build/demos/demo_comrv_baremetal.py:
```diff
@@ -27,8 +32,18 @@ def __init__(self):
+    ]
+
+    self.listDemoSpecificLinkerFlags = [
+      '-Wl,--grouping-file=' + self.strGrpFile,
```
add comment to explain -grouping-file. this is user flag, we want them to understand

ronen-haen-sndk replied:
done in caec55c

oferShinaar commented on WD-Firmware/demo/build/demos/demo_comrv_baremetal.py:
```diff
@@ -27,8 +32,18 @@ def __init__(self):
+
+    self.listDemoSpecificLinkerFlags = [
+      '-Wl,--grouping-file=' + self.strGrpFile,
+      '-Wl,--defsym=__comrv_cache_size=' + self.strComrvCacheSize
```
I guess this is due the ld issue. lets add todo here, hope this will go away. if not we need comment

ronen-haen-sndk replied:
done in caec55c

oferShinaar commented on WD-Firmware/demo/build/utils.py:
```diff
@@ -36,32 +36,97 @@
+         print ("Error: can't move .ovlgrpdata")
+         print "'%s' is too small [%s] while '%s' size is [%s]" %(STR_RESERVED_OVL_SEC_NAME, hex(intReservedSectionSize), STR_OVL_DATA_SEC_NAME, hex(intOvlSectionSize))
+         os.system("rm " + env['ELF_FILE'])
+   return None
```
there is strange ^M here

ronen-haen-sndk replied:
Done in e7e5bff

oferShinaar commented on WD-Firmware/demo/build/utils.py:
```diff
@@ -70,17 +135,54 @@ def fnCheckInstalledDependencis(listDependencis):
+       if not env['RISCV_LLVM_TC_PATH']:
+         print ("Error: Set environment variable '" + STR_RV_LLVM_ENV_VAR_NAME + "' to point to the RISCV llvm root")
+         exit(1)
+       # check if the binutils environment variable is set or empty
```
can you add one line space here (after exit)

ronen-haen-sndk replied:
done in eccffd7

oferShinaar commented on WD-Firmware/demo/demo_comrv_baremetal.c:
```diff
@@ -0,0 +1,179 @@
+}
+
+/* non overlay function */
+__attribute__((noinline)) void NonOverlayFunc(void)
```
use the D_xxx you defined

ronen-haen-sndk replied:
done in 6c72e9f

oferShinaar commented on WD-Firmware/demo/demo_comrv_baremetal.c:
```diff
@@ -0,0 +1,179 @@
+   comrvStatus_t stComrvStatus;
+   comrvGetStatus(&stComrvStatus);
+   /* we can't continue so loop forever */
+   while (1);
```
to illuminate compiler "purging" or optimization, please add "nops" inside the loop (can be new Macro for endless loop)

ronen-haen-sndk replied:
done in 6c72e9f

oferShinaar reviewed (changes requested):
most of the comments are minor. Just one main comment and that is the config of the ComRV that should drive from a file and not from the build system.
basically once this is done, this PL is approved

ronen-haen-sndk committed:
align code to new llvm:
- use .option nowarnreservedreg to disable warning of reserved regs usage
- add function pointer usage to the demo

ronen-haen-sndk committed:
add comments for the comrv related sections

ronen-haen-sndk committed:
CR changes:
add comments for disabling the build warning 'reserved registers use'

ronen-haen-sndk committed:
CR changes:
use psp defines for inline

ronen-haen-sndk committed:
CR changes:
add comment for _BUILTIN_EXPECT

ronen-haen-sndk committed:
CR changes:
move comrv defines from SConscript_comrv_baremetal to comrv_config.h

ronen-haen-sndk committed:
CR changes:
add comment

ronen-haen-sndk committed:
CR changes:beatify

ronen-haen-sndk committed:
CR changes:
add comments

ronen-haen-sndk committed:
CR changes:
remove ^M

ronen-haen-sndk committed:
CR changes: beautify

ronen-haen-sndk committed:
CR changes:
use a define for noinline
add a macro for endless loop

ronen-haen-sndk committed:
split linker script:
- temporary change to split linker script
- fix freertos build (caused by comrv change)

ronen-haen-sndk committed:
merge w/ comrv-master

ronen-haen-sndk committed:
fix build issue and align comrv cache size to this demo

Commit:
replacing toolchain environment variables with relative paths

Commit:
remove bin folder from toolchain and fix extract path of gcc

Commit:
adding LLVM and Binutils toolchain, updating the README file

ronen-haen-sndk committed:
change demo + fix comrv engine fail with group size > 512B

ronen-haen-sndk committed:
rename some demo macros

Commit:
adding symbolic link to bin folder in toolchain - GDB location

nidalf committed:
Merge branch 'comrv-master' into toolchain-changes

nidalf committed:
Merge pull request #21 from westerndigitalcorporation/toolchain-changes

replacing toolchain environment variables with relative paths

ronen-haen-sndk committed:
add rules

ronen-haen-sndk committed:
port comrv-baremetal demo to nexys

ronen-haen-sndk committed:
merge with master

oferShinaar committed:
move interupt registeration to demo. each demo will choose is interrupt vec method: baremetal or with rtos. NOTE NOTE NOTE NOTE: after merge with master the _init function for hifive1 will be changed, it will not longer set interrupt vect. for this commit the init is still the same as it was since comrv_master is not merged with master yet

Commit:
selecting llvm toolchain automatically if user selected comrv demo, fixing README

Commit:
Merge branch 'comrv-master' of https://github.com/westerndigitalcorporation/riscv-fw-infrastructure into comrv-master

ronen-haen-sndk committed:
remove bin/riscv64-unknown-elf-gdb

ronen-haen-sndk committed:
Merge branch 'comrv-master' of https://github.com/westerndigitalcorporation/riscv-fw-infrastructure into loc/comrv-master

Commit:
link all bin files from the toolchain to bin folder and rename all riscv32 to riscv64

ronen-haen-sndk committed:
Merge branch 'comrv-master' of https://github.com/westerndigitalcorporation/riscv-fw-infrastructure into loc/comrv-master

ronen-haen-sndk committed:
Merge remote-tracking branch 'origin/master' into loc/comrv-master
